lom/src/engine/sequenceHandler.go

```diff
@@ -0,0 +1,840 @@
+        LogError("Expect non null MsgSendServerResponse")
+    } else if msg.ReqType != TypeServerRequestAction {
+        LogError("Unexpected response req type (%d)/(%s)",
+                msg.ReqType, ServerReqTypeToStr[msg.ReqType])
+    } else {
+        m := msg.ResData
+        if data, ok := m.(ActionResponseData); !ok {
+            LogError("Unexpected response res data (ActionResponseData)/(%T)", m)
+        } else {
+            p.processActionResponse(&data)
+        }
+    }
+}
+
+
+/* Validate & drop from active requests */
+func (p *SeqHandler_t) validate_with_active(data *ActionResponseData, seq *sequenceState_t) error {
+    if r, ok := p.activeRequests[data.Action]; ok {
+        if (seq != nil) && (seq.currentRequest != r) {
+            return LogError("Response's req (%v) != current(%v)", r, seq.currentRequest)
+        }
+        if r.req.ReqType != TypeServerRequestAction  {
+            return LogError("Active requests type (%v) != TypeServerRequestAction", r)
+        }
+        rd := r.req.ReqData
+        if ar, ok1 := rd.(*ActionRequestData); !ok1 {
+            return LogError("Active requests data (%T) != ActionRequestData (%v)", rd, r)
+        } else if ar.InstanceId != data.InstanceId {
+            /* stale response */
+            return LogError("Active req vs res instance ID mismatch (%v) (%v)", ar, data)
+        }
+    } else {
+        return LogError("No matching req for res (%v)", data)
+    }
+    return nil
+}
+
+
+/*
+ * Called upon response received from client for an action.
+ *
+ * Validate response; publish it;
+ * validate corresponding request; Upon match, drop from active requests; else bail out.
+ *
+ * Find corresponding sequence by anomaly ID. 
+ * If not found, validate it for first action in sequence; else if found, validate
+ * against current request for this sequence. 
+ *
+ * if first action, but failed, re-publish with state as complete & skip creating
+ * sequence. Else if not first action, try resume current seq as this response could
+ * have been blocking it from running.
+ *
+ * If first action and succeeded, create a new sequence and add to pending seq list.
+ * Else if seq found, save the response in context. If response is success, call
+ * resumeSequence to kick off next request. Else call complete sequence.
+ * 
+ * For any error, the anomaly is re-published with state=complete and seq is completed,
+ * if appropriate.
+ * Try resumeNextSequence.
+ */
+func (p *SeqHandler_t) processActionResponse(data *ActionResponseData) {
+    if data == nil {
+        LogError("Internal error: Nil ActionResponseData")
+        return
+    }
+    if !data.Validate() {
+        LogError("Invalid ActionResponseData (%v)", *data)
+        return
+    }
+
+    anomalyID := data.AnomalyInstanceId
+    errCode := LoMResponseOk
+    errStr := ""
+    var err error = nil
+    seq, ok := p.sequencesByAnomaly[anomalyID]
+    if !ok {
+        seq = nil
+    } else if !seq.validate() {
```

Comment:
Is it needed ? When sequence is created outside this function, it should be good right ?